File: api/src/resource/question.rs
use api_derive::ApiResource;
use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};
use uuid::Uuid;

#[derive(ApiResource, Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct QuestionResource {
    pub id: Uuid,
    pub number: i32,
    pub author_id: Uuid,
    pub repository_id: Uuid,
    pub title: String,
    pub body: String,
    pub upvote: i32,
    pub impression: i32,
    pub created_at: DateTime<Utc>,
    pub updated_at: DateTime<Utc>,
    pub user_vote: Option<i16>,
    pub author: Option<AuthorResource>,
    pub comments: Vec<CommentResource>,
    pub answers: Vec<AnswerResource>,
}

#[derive(ApiResource, Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct AnswerResource {
    pub id: Uuid,
    pub question_id: Uuid,
    pub author_id: Uuid,
    pub body: String,
    pub upvote: i32,
    pub created_at: DateTime<Utc>,
    pub updated_at: DateTime<Utc>,
    pub user_vote: Option<i16>,
    pub author: Option<AuthorResource>,
    pub comments: Vec<CommentResource>,
}

#[derive(ApiResource, Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct CommentResource {
    pub id: Uuid,
    pub parent_id: Uuid,
    pub author_id: Uuid,
    pub body: String,
    pub upvote: i32,
    pub created_at: DateTime<Utc>,
    pub updated_at: DateTime<Utc>,
    pub user_vote: Option<i16>,
    pub author: Option<AuthorResource>,
}

#[derive(ApiResource, Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct AuthorResource {
    pub id: Uuid,
    pub name: String,
}

#[derive(ApiResource, Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct VoteResource {
    pub target_id: Uuid,
    pub score: i32,
    pub user_vote: Option<i16>,
}

File: gitdot-api-ts/src/resource/question.ts
import { z } from "zod";

export const AuthorResource = z.object({
  id: z.string().uuid(),
  name: z.string(),
});
export type AuthorResource = z.infer<typeof AuthorResource>;

export const CommentResource = z.object({
  id: z.string().uuid(),
  parent_id: z.string().uuid(),
  author_id: z.string().uuid(),
  body: z.string(),
  upvote: z.number().int(),
  created_at: z.string().datetime(),
  updated_at: z.string().datetime(),
  user_vote: z.number().int().nullable(),
  author: AuthorResource.nullable(),
});
export type CommentResource = z.infer<typeof CommentResource>;

export const AnswerResource = z.object({
  id: z.string().uuid(),
  question_id: z.string().uuid(),
  author_id: z.string().uuid(),
  body: z.string(),
  upvote: z.number().int(),
  created_at: z.string().datetime(),
  updated_at: z.string().datetime(),
  user_vote: z.number().int().nullable(),
  author: AuthorResource.nullable(),
  comments: z.array(CommentResource),
});
export type AnswerResource = z.infer<typeof AnswerResource>;

export const QuestionResource = z.object({
  id: z.string().uuid(),
  number: z.number().int(),
  author_id: z.string().uuid(),
  repository_id: z.string().uuid(),
  title: z.string(),
  body: z.string(),
  upvote: z.number().int(),
  impression: z.number().int(),
  created_at: z.string().datetime(),
  updated_at: z.string().datetime(),
  user_vote: z.number().int().nullable(),
  author: AuthorResource.nullable(),
  comments: z.array(CommentResource),
  answers: z.array(AnswerResource),
});
export type QuestionResource = z.infer<typeof QuestionResource>;

export const VoteResource = z.object({
  target_id: z.string().uuid(),
  score: z.number().int(),
  user_vote: z.number().int().nullable(),
});
export type VoteResource = z.infer<typeof VoteResource>;
